### PR TITLE
#236 - created PrecedingMetadata with implementation

### PR DESCRIPTION
**Pull request by olenagerasimova**
Part of #236: created abstraction to check whether metadata file exists and to unpack it. This abstraction will be passed to `ModifiableMetadata` instead of `Path` to previous metadata file.

### CODE REVIEW

**Comment by olenagerasimova** [comment]
@0crat out

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/artipie/rpm-adapter/issues/325#issuecomment-654411822))

@olenagerasimova Job `gh:artipie/rpm-adapter#325` is not assigned, can't get performer

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/artipie/rpm-adapter/issues/325#issuecomment-654411822))

@olenagerasimova Pull request [#325](https://github.com/artipie/rpm-adapter/issues/325) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

File: src/test/java/com/artipie/rpm/pkg/PrecedingMetadataFromDirTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.cactoos.list.ListOf;
+import org.hamcrest.Matcher;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.AllOf;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.llorllale.cactoos.matchers.MatcherOf;
+
+/**
+ * Test for {@link PrecedingMetadata}.
```
Comment (olegmoz) [inline src/test/java/com/artipie/rpm/pkg/PrecedingMetadataFromDirTest.java:40]: @olenagerasimova tests are for `PrecedingMetadata.FromDir` class

File: src/test/java/com/artipie/rpm/pkg/PrecedingMetadataFromDirTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        MatcherAssert.assertThat(
+            new PrecedingMetadata.FromDir(pckg, temp),
+            new AllOf<PrecedingMetadata.FromDir>(
+                new ListOf<Matcher<? super PrecedingMetadata.FromDir>>(
+                    new MatcherOf<PrecedingMetadata.FromDir>(
+                        PrecedingMetadata.FromDir::exists
+                    ),
+                    new MatcherOf<PrecedingMetadata.FromDir>(
+                        meta -> { return meta.findAndUnzip().isPresent(); }
+                    )
```
Comment (olegmoz) [inline src/test/java/com/artipie/rpm/pkg/PrecedingMetadataFromDirTest.java:62]: @olenagerasimova I'd suggest to extract check of `findAndUnzip()` to another test. Right now this test method looks too massive and lacks verification of `Path` value returned by the method

File: src/main/java/com/artipie/rpm/pkg/PrecedingMetadata.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            final Optional<Path> res;
+            if (metadata.isPresent()) {
+                final Path unziped = Files.createTempFile(
+                    this.dir, String.format("old-%s", this.type.filename()), ".xml"
+                );
+                new Gzip(metadata.get()).unpack(unziped);
+                res = Optional.of(unziped);
+            } else {
+                res = Optional.empty();
+            }
```
Comment (olegmoz) [inline src/main/java/com/artipie/rpm/pkg/PrecedingMetadata.java:97]: @olenagerasimova it seems that this if/else block may be replaced by `this.find().map(...)`

Comment (olenagerasimova) [reply src/main/java/com/artipie/rpm/pkg/PrecedingMetadata.java:97]: @olegmoz I'd prefer to keep if/else here

**Comment by olenagerasimova** [comment]
@olegmoz corrections made, thanks

**Comment by 0crat** [comment]
Job `gh:artipie/rpm-adapter#325` is not assigned, can't get performer

**Pull request merged**